modules/configuration-annotation-processor/src/test/java/org/apache/ignite/configuration/internal/validation/ValidationUtilTest.java

```diff
@@ -0,0 +1,174 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.configuration.internal.validation;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.ignite.configuration.annotation.Config;
+import org.apache.ignite.configuration.annotation.ConfigValue;
+import org.apache.ignite.configuration.annotation.ConfigurationRoot;
+import org.apache.ignite.configuration.annotation.NamedConfigValue;
+import org.apache.ignite.configuration.annotation.Value;
+import org.apache.ignite.configuration.internal.RootsNode;
+import org.apache.ignite.configuration.internal.util.ConfigurationUtil;
+import org.apache.ignite.configuration.sample.storage.TestConfigurationStorage;
+import org.apache.ignite.configuration.tree.NamedListView;
+import org.apache.ignite.configuration.validation.ValidationContext;
+import org.apache.ignite.configuration.validation.ValidationIssue;
+import org.apache.ignite.configuration.validation.Validator;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.emptySet;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/** */
+public class ValidationUtilTest {
+    /** */
+    @Target(FIELD)
+    @Retention(RUNTIME)
+    @interface LeafValidation {
+    }
+
+    /** */
+    @Target(FIELD)
+    @Retention(RUNTIME)
+    @interface InnerValidation {
+    }
+
+    /** */
+    @Target(FIELD)
+    @Retention(RUNTIME)
+    @interface NamedListValidation {
+    }
+
+    /** */
+    @ConfigurationRoot(rootName = "root", storage = TestConfigurationStorage.class)
+    public static class ValidatedRootConfigurationSchema {
+        /** */
+        @InnerValidation
+        @ConfigValue
+        public ValidatedChildConfigurationSchema child;
+
+        /** */
+        @NamedListValidation
+        @NamedConfigValue
+        public ValidatedChildConfigurationSchema elements;
+    }
+
+    /** */
+    @Config
+    public static class ValidatedChildConfigurationSchema {
+        /** */
+        @LeafValidation
+        @Value(hasDefault = true)
```

Comment:
What do you think about renaming `hasDefault` to `withDefault`?

Reply:
Word "with" implies that expression will be followed by that very default, not by an indicator flag. So I think that "has" fits better.